docs/framework/react/guides/custom-errors.md

```diff
@@ -0,0 +1,267 @@
+---
+id: custom-errors
+title: Custom Errors
+---
+
+
+
+
+## Return String Values from Forms
+
+By default, TanStack Form allows you to return string error messages from validators:
+
+```tsx
+<form.Field
+  name="username"
+  validators={{
+    onChange: ({ value }) =>
+      value.length < 3 ? "Username must be at least 3 characters" : undefined
+  }}
+/>
+```
+
+For form-level validation affecting multiple fields:
+
+```tsx
+const form = useForm({
+  defaultValues: {
+    username: "",
+    email: ""
+  },
+  validators: {
+    onChange: ({ value }) => {
+      return {
+        fields: {
+          username: value.username.length < 3 ? "Username too short" : undefined,
+          email: !value.email.includes("@") ? "Invalid email" : undefined
+        }
+      };
+    }
+  }
+});
+```
+
+String errors are the most common type and are easily displayed in your UI:
+
+```tsx
+{field.state.meta.errors.map((error, i) => (
+  <div key={i} className="error">{error}</div>
+))}
+```
+
+## Return Non-String Values from Forms
+
+TanStack Form supports returning non-string values as errors, providing flexibility for different validation scenarios:
+
+### Numbers
+
+Useful for representing quantities, thresholds, or magnitudes:
+
+```tsx
+<form.Field
+  name="age"
+  validators={{
+    onChange: ({ value }) => value < 18 ? 18 - value : undefined
+  }}
+/>
+```
+
+Display in UI:
+```tsx
+{typeof field.state.meta.errors[0] === "number" && (
+  <div className="error">
+    You need {field.state.meta.errors[0]} more years to be eligible
+  </div>
+)}
+```
+
+### Booleans
+
+Simple flags to indicate error state:
+
+```tsx
+<form.Field
+  name="accepted"
+  validators={{
+    onChange: ({ value }) => !value ? true : undefined
+  }}
+/>
+```
+
+Display in UI:
+```tsx
+{field.state.meta.errors[0] === true && (
+  <div className="error">You must accept the terms</div>
+)}
+```
+
+### Objects
+
+Rich error objects with multiple properties:
+
+```tsx
+<form.Field
+  name="email"
+  validators={{
+    onChange: ({ value }) => {
+      if (!value.includes("@")) {
+        return {
+          message: "Invalid email format",
+          severity: "error",
+          code: 1001
+        };
+      }
+      return undefined;
+    }
+  }}
+/>
+```
+
+Display in UI:
+```tsx
+{typeof field.state.meta.errors[0] === "object" && (
+  <div className={`error ${field.state.meta.errors[0].severity}`}>
+    {field.state.meta.errors[0].message}
+    <small> (Code: {field.state.meta.errors[0].code})</small>
+  </div>
+)}
+```
+
+or
+```tsx
+{field.state.meta.errorMap.onChange?.message ? (
+    <p>{field.state.meta.errorMap.onBlur?.message}</p>
+  ) : null}
+```
+
+in the example above it depends on the event error you want to display either `onChange` `onBlur` or else, but of course with TypeScript support helping you out
```

Comment:
I wouldn't mix onChange and onBlur in the same snippet this way. Probably better to stick with one and if you want you can mention that other keys might exist in the errorMap, but should is defined somewhere else in the docs